Add limits to the number of ImageBuffers that the GPU process can create

heycam opened the pull request:
#### 80225aaf7451c18e20832dec77948e514d39d368
<pre>
Add limits to the number of ImageBuffers that the GPU process can create
<a href="https://bugs.webkit.org/show_bug.cgi?id=273326">https://bugs.webkit.org/show_bug.cgi?id=273326</a>
<a href="https://rdar.apple.com/95930955">rdar://95930955</a>

Reviewed by Matt Woodrow.

On Cocoa ports, ImageBuffers created in the GPU process consume both
global and process-wide resources. Accelerated ImageBuffers consume
IOSurface handles (which the kernel has a limited number of, and which
also have a per process limit) and Metal objects (which are also limited
per process). Unaccelerated ImageBuffers consume Mach port handles for
their ShareableBitmap handles (which are per process).

Prevent leaks (both undiagnosed and from poorly behaving Web content
creating 2D canvases) from exhausting these resources by placing limits
on the number of ImageBuffers the GPU process can create.

These limits are:

* 200,000 ImageBuffers for the GPU process (a number a bit lower than
  the 256K Mach port limit)
* 50,000 canvas ImageBuffers for each Web Content process (to prevent
  one Web Content process from using the entire global limit)
* 10,000 accelerated ImageBuffers for the GPU process (a number a bit
  lower than the 16K per-process IOSurface limit)
* 5,000 accelerated canvas ImageBuffers for each Web Content process (to
  prevent one Web Content process from using the entire global limit)

* LayoutTests/fast/canvas/image-buffer-resource-limits-expected.txt: Added.
* LayoutTests/fast/canvas/image-buffer-resource-limits.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::getEffectiveRenderingModeForTesting):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::getImageBufferResourceLimitsForTesting const):
* Source/WebCore/platform/graphics/ImageBufferResourceLimits.h: Added.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::getEffectiveRenderingModeOfNewlyCreatedAcceleratedImageBuffer):
(WebCore::Internals::getImageBufferResourceLimits):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/RemoteSharedResourceCache.cpp:
(WebKit::RemoteSharedResourceCache::didCreateImageBuffer):
(WebKit::RemoteSharedResourceCache::didReleaseImageBuffer):
(WebKit::RemoteSharedResourceCache::reachedAcceleratedImageBufferLimit const):
(WebKit::RemoteSharedResourceCache::reachedImageBufferForCanvasLimit const):
(WebKit::RemoteSharedResourceCache::getResourceLimitsForTesting const):
* Source/WebKit/GPUProcess/RemoteSharedResourceCache.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::RemoteImageBuffer):
(WebKit::RemoteImageBuffer::~RemoteImageBuffer):
(WebKit::RemoteImageBuffer::stopListeningForIPC):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::adjustImageBufferRenderingMode):
(WebKit::RemoteRenderingBackend::allocateImageBuffer):
(WebKit::RemoteRenderingBackend::getImageBufferResourceLimitsForTesting):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::sendWithAsyncReply):
(WebKit::RemoteRenderingBackendProxy::getImageBufferResourceLimitsForTesting):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
(WebKit::RemoteRenderingBackendProxy::sendWithAsyncReply):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::getImageBufferResourceLimitsForTesting const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:

Canonical link: <a href="https://commits.webkit.org/281395@main">https://commits.webkit.org/281395@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1af0dde7ac64f5e9ecf3ed5ecfc2e4a23851d548

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59740 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39086 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12268 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63656 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10264 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61869 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46738 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10443 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48448 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7177 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61770 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36469 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51726 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29288 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33175 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8969 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9187 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55109 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9248 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65387 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3668 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9133 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55790 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3679 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51710 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55926 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3049 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8933 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34899 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35982 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37068 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35727 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebKit/GPUProcess/RemoteSharedResourceCache.cpp:
```diff
@@ -74,6 +89,54 @@ void RemoteSharedResourceCache::lowMemoryHandler()
+ImageBufferResourceLimits RemoteSharedResourceCache::getResourceLimitsForTesting() const
+{
+    return {
+        .acceleratedImageBufferForCanvasCount = m_acceleratedImageBufferForCanvasCount,
```
I had not seen this syntax before, and I really like it.

heycam committed:
Add limits to the number of ImageBuffers that the GPU process can create
https://bugs.webkit.org/show_bug.cgi?id=273326
rdar://95930955

Reviewed by Matt Woodrow.

On Cocoa ports, ImageBuffers created in the GPU process consume both
global and process-wide resources. Accelerated ImageBuffers consume
IOSurface handles (which the kernel has a limited number of, and which
also have a per process limit) and Metal objects (which are also limited
per process). Unaccelerated ImageBuffers consume Mach port handles for
their ShareableBitmap handles (which are per process).

Prevent leaks (both undiagnosed and from poorly behaving Web content
creating 2D canvases) from exhausting these resources by placing limits
on the number of ImageBuffers the GPU process can create.

These limits are:

* 200,000 ImageBuffers for the GPU process (a number a bit lower than
  the 256K Mach port limit)
* 50,000 canvas ImageBuffers for each Web Content process (to prevent
  one Web Content process from using the entire global limit)
* 10,000 accelerated ImageBuffers for the GPU process (a number a bit
  lower than the 16K per-process IOSurface limit)
* 5,000 accelerated canvas ImageBuffers for each Web Content process (to
  prevent one Web Content process from using the entire global limit)

* LayoutTests/fast/canvas/image-buffer-resource-limits-expected.txt: Added.
* LayoutTests/fast/canvas/image-buffer-resource-limits.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::getEffectiveRenderingModeForTesting):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::getImageBufferResourceLimitsForTesting const):
* Source/WebCore/platform/graphics/ImageBufferResourceLimits.h: Added.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::getEffectiveRenderingModeOfNewlyCreatedAcceleratedImageBuffer):
(WebCore::Internals::getImageBufferResourceLimits):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/RemoteSharedResourceCache.cpp:
(WebKit::RemoteSharedResourceCache::didCreateImageBuffer):
(WebKit::RemoteSharedResourceCache::didReleaseImageBuffer):
(WebKit::RemoteSharedResourceCache::reachedAcceleratedImageBufferLimit const):
(WebKit::RemoteSharedResourceCache::reachedImageBufferForCanvasLimit const):
(WebKit::RemoteSharedResourceCache::getResourceLimitsForTesting const):
* Source/WebKit/GPUProcess/RemoteSharedResourceCache.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::RemoteImageBuffer):
(WebKit::RemoteImageBuffer::~RemoteImageBuffer):
(WebKit::RemoteImageBuffer::stopListeningForIPC):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::adjustImageBufferRenderingMode):
(WebKit::RemoteRenderingBackend::allocateImageBuffer):
(WebKit::RemoteRenderingBackend::getImageBufferResourceLimitsForTesting):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::sendWithAsyncReply):
(WebKit::RemoteRenderingBackendProxy::getImageBufferResourceLimitsForTesting):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
(WebKit::RemoteRenderingBackendProxy::sendWithAsyncReply):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::getImageBufferResourceLimitsForTesting const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:

Canonical link: https://commits.webkit.org/281395@main
